00-programming-fundamentals/intro-to-regex.md

```diff
@@ -0,0 +1,133 @@
+
+So `/a(ab)*a/` would match the strings:  "aaba", "aa", "aababa" because it requires an a followed by zero or more "ab" blocks and ending with an "a".
+
+Similarly, `/\$(\d)+\.(\d){2}` would match us currency, like you were asked to deduce above requiring a dollar sign followed by one or more digits, followed by a period and ending with 2 digits.  
+
+
+
+### Practice
+
+See if you can write a regular expression for these patterns:
+
+1. A 5 digit number
+1. An IP address like 127.0.0.1, or 206.190.36.45
+
+[Sample Solutions](https://gist.github.com/CheezItMan/54e39aa4a355aa5a24bb281d5d360121)
+
+## List of Regex Special Characters
+
+| Character | Name  |  Description |
+| --- | --- | --- |
+|   \	|   Backslash	|   The backslash gives special meaning to the character following it. For example, the combination "\n" stands for the newline, one of the control characters. 	|
+|   ^	|   Caret	|   The caret is the start of line anchor or the negate symbol.  Example: "^a" matches "a" at the start of a line.  Example: "[^0-9]" matches any non digit.	|
+|   $	|   Dollar	|   $ the dollar is the end of line anchor.	|
+|   {}	|   Curly Braces	|   { } the open and close curly bracket are used as range quantifiers.	|
+|   []	|   Square Brackets	|    Open and close square bracket define a character class to match a single character.	|
+|   ()	|   Parentheses 	|   The open and close parenthesis are used for grouping characters	|
+|   .	|   Dot	|   the dot matches any character except the newline.	|
+|   *	|   Star	|   The star is the match-zero-or-more quantifier.	|
+|   +	|   Plus	|   The plus is the match-zero-or-one quantifier.	|
+|   ?	|   Question Mark	|   The question mark	 is the match-one-or-more quantifier.	|
+|   &#124;	|   Pipe	|   The vertical pipe separates a series of alternatives.	|
+|  < > 	|   Anchors	|   The smaller and greater signs are anchors that specify a left or right word boundary.	|
+|  - 	|   Minus	|   the minus indicates a range in a character class (when it is not at the first position after the "[" opening bracket or the last position before the "]" closing bracket. For example "[A-Z]" matches any uppercase character.	|
+| & | Ampersand | The and is the "substitute complete match" symbol. |
+
+
+
+## Resources
+-  [Regex in Wikipedia](https://en.wikipedia.org/wiki/Regular_expression)
+-  [Regular Expressions in rubylearning.com](http://rubylearning.com/satishtalim/ruby_regular_expressions.html)
+-  [Regexpal](http://regexpal.com.s3-website-us-east-1.amazonaws.com/?_ga=2.249565263.813004562.1495485737-848749570.1493938714) - A useful tool for composing Regular Expressions
+- [Regexper](https://regexper.com/) - a fantastic tool to generate a visual diagram of a regular expression.
```

Comment:
I think it would be good to generate some of these diagrams and add them to the notes throughout since some folks will see this and gain a much better comprehension of the material